### PR TITLE
Rk/split decimal state into smaller states nullable long state

### PR DESCRIPTION
**Pull request by radek-starburst**

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 1]
```diff
@@ -114,119 +119,131 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
...
+                decimalOffset);
 
-        state.addOverflow(overflow);
+        overflow += overflowState.getValue();
+        overflowState.setNull(overflow == 0 & overflowState.isNull());
+        overflowState.setValue(overflow);
     }
 
-    public static void combine(LongDecimalWithOverflowAndLongState state, LongDecimalWithOverflowAndLongState otherState)
+    public static void combine(Int128State decimalState, LongState counterState, NullableLongState overflowState, Int128State otherDecimalState, LongState otherCounterState, NullableLongState otherOverflowState)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:181]: you should use annotated functions like in `io.trino.operator.aggregation.AverageAggregations`

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:181]: you should base your code of these two commits:
https://github.com/dain/trino/commit/77365dbb5bfb14c69fb14d82bcf15001067dcfae
https://github.com/dain/trino/commit/a86e3840cc3e046dd764fa56a06e128461d2d78a

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:181]: We want annotated aggregations (I missed that during initial code skimming). I don't think https://github.com/trinodb/trino/pull/11477 should be blocked on this PR as this PR requires investigation

**Review by skrzypo987 — COMMENTED** [review]
I skimmed it and don't see any obvious performance bugs.
You need to choose few queries that regress the most and run benchmark on them with jfr profiling. Both before and after. That should bring some answers

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 2]
```diff
@@ -114,119 +119,131 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
...
+                decimal[decimalOffset + 1],
                 rightHigh,
                 rightLow,
                 decimal,
-                offset);
+                decimalOffset);
 
-        state.addOverflow(overflow);
+        overflow += overflowState.getValue();
+        overflowState.setNull(overflow == 0 & overflowState.isNull());
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:153]: why not just `overflowState.setNull(overflow == 0)`?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 3]
```diff
@@ -114,119 +119,131 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
...
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:177]: why not just `overflowState.setNull(overflow == 0)`?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java [hunk 4]
```diff
@@ -95,91 +95,101 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
...
+                decimal[decimalOffset + 1],
                 rightHigh,
                 rightLow,
                 decimal,
-                offset);
-        state.setOverflow(Math.addExact(overflow, state.getOverflow()));
+                decimalOffset);
+        overflow = Math.addExact(overflow, overflowState.getValue());
+        overflowState.setValue(overflow);
+        overflowState.setNull(overflow == 0 & overflowState.isNull());
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java:127]: same comments as above `overflowState.setNull(overflow == 0)`

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java [hunk 5]
```diff
@@ -95,91 +95,101 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
...
+                    decimal,
+                    decimalOffset);
+        } else {
+            decimal[decimalOffset] = otherDecimal[otherDecimalOffset];
+            decimal[decimalOffset + 1] = otherDecimal[otherDecimalOffset + 1];
         }
+        decimalState.setNotNull(true);
+        overflow = (overflowState.isNull() ? 0 : overflowState.getValue()) + Math.addExact(overflow, otherOverflowState.isNull() ? 0 : otherOverflowState.getValue());
+        overflowState.setValue(overflow);
+        overflowState.setNull(overflow == 0 & overflowState.isNull() & otherOverflowState.isNull());
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java:178]: ditto

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/Int128State.java [hunk 6]
```diff
@@ -16,21 +16,17 @@
...
+public interface Int128State
         extends AccumulatorState
 {
     boolean isNotNull();
 
-    void setNotNull();
+    void setNotNull(boolean notNull);
 
-    long[] getDecimalArray();
+    void addNotNull(boolean isNotNull);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/Int128State.java:27]: unused, remove

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 7]
```diff
@@ -114,119 +119,131 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
...
+                decimal[decimalOffset],
+                decimal[decimalOffset + 1],
                 rightHigh,
                 rightLow,
                 decimal,
-                offset);
+                decimalOffset);
 
-        state.addOverflow(overflow);
+        overflow += overflowState.getValue();
```
[the remark this answers is not in the document]
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:152]: I can do it but I did it before and it has no impact on performance (I mean, it was checked).
Call to `getValue` is inlined so and as a result the code is +/- same as we use `addValue`.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java (hunk 2 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:153]: It is not necessary. It was done to do not change the previouse behaviour. Generally, it does not matter but you right.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java (hunk 3 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:177]: It is not necessary. It was done to do not change the previouse behaviour. Generally, it does not matter but you right.

**Pull request closed without merging**